File: src/main/java/com/github/blueminecraftteam/healthmod/statuseffects/WoundInfectionStatusEffect.java
/*
 * Copyright (c) 2020 Blue Minecraft Team.
 *
 * This file is part of HealthMod.
 *
 * HealthMod is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * HealthMod is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with HealthMod.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.blueminecraftteam.healthmod.statuseffects;

import com.github.blueminecraftteam.healthmod.mixins.DamageSourceAccessorMixin;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectType;

public class WoundInfectionStatusEffect extends StatusEffect {
    public WoundInfectionStatusEffect(StatusEffectType type, int color) {
        super(type, color);
    }

    @Override
    public void applyUpdateEffect(LivingEntity entity, int amplifier) {
        entity.damage(DamageSourceAccessorMixin.newDamageSource("wound_infection"), 2.5F);
    }

    @Override
    public boolean canApplyUpdateEffect(int duration, int amplifier) {
        int k = 25 >> amplifier;
        if (k > 0) {
            return duration % k == 0;
        } else {
            return true;
        }
    }
}

File: src/main/kotlin/io/github/blueminecraftteam/healthmod/statuseffects/WoundInfectionStatusEffect.kt
/*
 * Copyright (c) 2020 Blue Minecraft Team.
 *
 * This file is part of HealthMod.
 *
 * HealthMod is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * HealthMod is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with HealthMod.  If not, see <https://www.gnu.org/licenses/>.
 */
package io.github.blueminecraftteam.healthmod.statuseffects

import io.github.blueminecraftteam.healthmod.mixin.DamageSourceAccessorMixin
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.effect.StatusEffect
import net.minecraft.entity.effect.StatusEffectType

class WoundInfectionStatusEffect(type: StatusEffectType, color: Int) : StatusEffect(type, color) {
    override fun applyUpdateEffect(entity: LivingEntity, amplifier: Int) {
        entity.damage(DAMAGE_SOURCE, 2.5f)
    }

    override fun canApplyUpdateEffect(duration: Int, amplifier: Int): Boolean {
        val k = 25 shr amplifier

        return if (k > 0) duration % k == 0 else true
    }

    companion object {
        private val DAMAGE_SOURCE = DamageSourceAccessorMixin.newDamageSource("wound_infection")
    }
}